ros2_batch_job/__main__.py

```diff
@@ -356,6 +371,51 @@ def build_and_test(args, job):
     # return 0 if ret_test == 0 and ret_testr == 0 else 1
     return 0
 
+def build_with_sanitizer(cmd, args, job):
+    print('# BEGIN SUBSECTION: build with sanitizer')
+    sanitizer_type = args.enable_sanitizer_type
+    sanitizer_cmake_args = [
+    '-DCMAKE_CXX_FLAGS="-fsanitize=' + sanitizer_type + ' -g"',
+    '-DCMAKE_C_FLAGS="-fsanitize=' + sanitizer_type + ' -g"',
+    '-DCMAKE_EXE_LINKER_FLAGS="-fsanitize=' + sanitizer_type + '"',
+    '-DCMAKE_MODULE_LINKER_FLAGS="-fsanitize=' + sanitizer_type + '"',
+    '-DCMAKE_SHARED_LINKER_FLAGS="-fsanitize=' + sanitizer_type + '"'
+    ]
+    index = cmd.index('--cmake-args')
+    cmd[index + 1:index + 1] = sanitizer_cmake_args
+    if args.restrict_san_pkgs_regex:
+        cmd.extend(['--packages-select-regex ' + args.restrict_san_pkgs_regex])
+    print(cmd)
+    ret_build = job.run(cmd, shell=True)
+    info("colcon build with sanitizer returned: '{0}'".format(ret_build))
+    print('# END SUBSECTION')
+    return ret_build
+
+def update_test_cmd_based_on_sanitizer_args(test_cmd, args):
+    prefix_test_cmd = get_prefix_test_cmd_based_on_sanitizer_args(args)
+    suffix_test_cmd = get_suffix_test_cmd_based_on_sanitizer_args(args)
+    return prefix_test_cmd + test_cmd + suffix_test_cmd
+
+def get_prefix_test_cmd_based_on_sanitizer_args(args):
+    export_library_preload_for_sanitizer = []
+    if args.enable_sanitizer_type == 'address':
+        # Address Sanitizer runtime needs to be the first in the initial loaded library list
+        export_library_preload_for_sanitizer = ['export LD_PRELOAD=$(ls /usr/lib/x86_64-linux-gnu/libasan.so* | head -1);']
```

Comment:
Let's use /usr/lib/x86_64-linux-gnu/libasan.so and drop the | head -n 1 - we can do exact matching here and it will less fragile (or not less fragile than randomly selecting the first versioned shared library which may or may not match what's actually got linked with the binary).

Reply:
This is a bit tricky as there is no generic libasan.so. I can only see libasan.so.4 and libasan.so.4.0.0. I will try to see if there is any other way to identify the path. One naive way to do it might be to run ld on a ros2 generated binary and grepping for libasan, not sure how to do it though